### PR TITLE
Refactor CSS

### PR DESCRIPTION
**Pull request by heyitsmeuralex**
See #185. This doesn't actually add a dark theme, but gets us 85% of the way there. Work-in-progress.

~~Blocked by #225 as this includes it.~~

### CODE REVIEW

**Comment by heyitsmeuralex** [comment]
Requested a review from @towerofnix as the rest of the work is just refactoring. Would like an opinion on the style etc.

**Comment by heyitsmeuralex** [comment]
Only thing left is to make more prescriptive variables as discussed in #185.

File: packages/client/css/app.css [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+a:any-link.link, .Link {
+  color: var(--blue);
+  text-decoration: underline;
+}
+
+::-moz-selection { background: #b3d4fc; }
+::selection { background: #b3d4fc; }
+
+table, th, td {
```
Comment (PullJosh) [inline packages/client/css/app.css]: Feeling a little nervous about such generic selectors. Which tables are these styles meant to apply to? Where do they appear?

File: packages/client/css/generic/button.css [hunk 2]
```diff
@@ -0,0 +1,28 @@
+.Button {
+  padding: 10px 16px;
+  &:not(:first-child) { margin-left: 3px; }
+  &:not(:last-child) { margin-right: 3px; }
```
Comment (PullJosh) [inline packages/client/css/generic/button.css]: Adding negatives (`:not`) tends to make things more confusing. Might it be better to say the following?

```css
margin-right: 6px;
&:last-child { margin-right: 0; }
```

This, I believe, would achieve the same effect: a `6px` gap between each `.Button`.

File: packages/client/css/generic/input.css [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    font-weight: bold;
+    font-size: 16px;
+
+    text-transform: uppercase;
+    margin-bottom: 4px;
+    margin-left: 13px;
+
+    -webkit-touch-callout: none;
+    -webkit-user-select: none;
```
Comment (PullJosh) [inline packages/client/css/generic/input.css]: Indentation doesn't quite line up. This problem probably also existed in the old CSS, but it should be fixed while we're in the process of reworking things.

File: packages/client/css/generic/input.css [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  background: var(--gray-900);
+
+  border: 0;
+  border-bottom: 1px solid var(--gray-500);
+  &:focus { border-color: var(--blue); }
+
+  &[disabled] {
+    cursor: not-allowed;
+    background: var(--gray-700);
+    opacity: .8;
```
Comment (PullJosh) [inline packages/client/css/generic/input.css]: There seems to be some inconsistency with whether or not to include a leading 0. We should choose a standard (my vote is for including the 0) and mention it in CONTRIBUTING.md.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -75,3 +75,77 @@
...
+    &.is-loading {
+      /* ... */
+    }
+  }
+</style>
+
+<div class='Image is-loading'></div>
+```
+
+* Use `&-childName` for children
```
Comment (PullJosh) [inline CONTRIBUTING.md]: It seems that child component names begin with a lowercase, while parent components begin with an uppercase. Is this intentional? More importantly, is it a good idea?

File: packages/client/css/generic/message-group.css [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+.MessageGroup-authorUsername {
+  display: inline-block;
+  margin-right: 4px;
+
+  font-weight: bold;
+  text-align: right;
+}
+
+.MessageGroup-contents {
+  width: calc(100% - 48px - 16px);
```
Comment (PullJosh) [inline packages/client/css/generic/message-group.css]: What is the reasoning for these numbers? I love separating out `48px` and `16px` (as opposed to `100% - 32px`) if they have distinct meanings, but it's not clear from the code what those meanings are.

A simple comment should clear things up.

File: packages/client/css/generic/message.css [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  object-fit: contain;
+  width: 1rem;
+
+  margin: 0 .05em 0 .1em;
+  vertical-align: -.4em;
+}
+
+/* note: this animation name is used in the JavaScript code,
+   so take care if renaming it */
```
Comment (PullJosh) [inline packages/client/css/generic/message.css]: :+1: This comment will save somebody one day.

File: packages/client/css/pages/account-settings.css [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    }
+  }
+
+.AccountSettings-submit {
+  display: block;
+
+  margin-top: 80px;
+  margin-left: auto;
+  margin-right: 0;
+  margin-bottom: 20px;
```
Comment (PullJosh) [inline packages/client/css/pages/account-settings.css]: Any reason not to consolidate these into a single `margin`?

File: packages/client/css/specific/modal.css [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  left: 50%;
+  transform: translate(-50%, -50%);
+  max-width: calc(100% - 48px);
+
+  background: var(--gray-900);
+  border-radius: 4px;
+  overflow: hidden;
+
+  z-index: 950;
+  box-shadow: 0 1px 8px #00000044;
```
Comment (PullJosh) [inline packages/client/css/specific/modal.css]: 8-digit hex notation [doesn't seem](https://caniuse.com/#feat=css-rrggbbaa) particularly well supported. I know that browser support isn't priority numero uno, but it seems no worse to use `rgba(0, 0, 0, 0.27)` in this case.

File: packages/client/css/specific/modal.css [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  top: 16px;
+  right: 16px;
+  width: 19px;
+  height: 19px;
+  background: url('/img/x.svg') center center / contain;
+  cursor: pointer;
+}
+
+.Modal-title {
+  padding: 69px;
```
Comment (PullJosh) [inline packages/client/css/specific/modal.css]: classic

File: packages/client/css/app.css (hunk 1 above)
Comment (towerofnix) [reply packages/client/css/app.css]: The only tables in use are currently in the emote and authorized users setting pages, I think.

File: packages/client/css/generic/input.css (hunk 4 above)
Comment (towerofnix) [reply packages/client/css/generic/input.css]: My vote is 1000% for including the zero.

File: packages/client/css/generic/message-group.css (hunk 6 above)
Comment (towerofnix) [reply packages/client/css/generic/message-group.css]: Gosh, yeah. @heyitsmeuralex I absolutely can't stand using `calc`. Could you rework this with flexbox, or even just float?

File: packages/client/css/pages/account-settings.css (hunk 8 above)
Comment (towerofnix) [reply packages/client/css/pages/account-settings.css]: (Note the order would be `margin: 80px 0 20px auto`.)

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -75,3 +75,77 @@
 +   3,
   ]
 ```
+
+## CSS Naming Conventions
+
+This is essentially our own flavour of BEM.
+
+* Use `.ComponentName` for components
```
[the remark this answers is not in the document]
Comment (heyitsmeuralex) [reply CONTRIBUTING.md]: Yeah. `Use pascal-case...`

File: packages/client/css/app.css (hunk 1 above)
Comment (heyitsmeuralex) [reply packages/client/css/app.css]: Yeah, these are leftovers! We should move them to more specific places.

File: packages/client/css/generic/input.css (hunk 4 above)
Comment (heyitsmeuralex) [reply packages/client/css/generic/input.css]: wrong

File: CONTRIBUTING.md (hunk 5 above)
Comment (heyitsmeuralex) [reply CONTRIBUTING.md]: Yeah, it's intentional. I think it makes it more obvious what the parent and child are.

File: packages/client/css/generic/message-group.css (hunk 6 above)
Comment (heyitsmeuralex) [reply packages/client/css/generic/message-group.css]: aaaa No idea what this line is for :package:
I'll look into it.

File: packages/client/css/generic/message.css (hunk 7 above)
Comment (heyitsmeuralex) [reply packages/client/css/generic/message.css]: `git blame` @towerofnix :star:

**Comment by heyitsmeuralex** [comment]
Made all those requested changes, @PullJosh. Only thing left is `--prescriptive-variable-names`.

**Comment by heyitsmeuralex** [comment]
Made all those requested changes, @PullJosh. Only thing left is `--prescriptive-variable-names`.

**Pull request merged**